pkg/toystore/destroyguard/guard.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package destroyguard
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/cockroachdb/cockroach/pkg/util"
+)
+
+// A Guard tracks a reference count (usually associated to some object to be
+// guarded) while offering an API that allows tearing down the guard and
+// instructing all reference holders to relinquish their references in a timely
+// manner.
+//
+// A Guard is in either of three states, through which it moves in the following
+// order:
+// - active: Teardown has not been called; Acquire() returns success.
+// - tearing down: Teardown has been called, but the refcount hasn't dropped to
+//   zero; Acquire returns the error passed to Teardown.
+// - deadCh: Teardown has been called and the refcount is zero; the Dead() channel
+//   is closed and Acquire continues to return the same error.
+//
+// The zero Guard is ready to use. A Guard must not be copied by value. All
+// methods on Guard are safe to call concurrently.
+type Guard struct {
+	// TODO(tbg): use int32 here to assert Release() without Acquire().
+	ref uint32 // updated atomically by Acquire() and Release()
```

Comment:
I'd like for us to start using a naming convention for things that must be accessed using `atomic`, so it's clear without reading the comment on the field when it's being used incorrectly (similar to the `mu struct { syncutil.Mutex ... }` pattern we use elsewhere). Thoughts?

The particular convention I've tried in a couple places is an "Atomic" suffix, but I'm not tied to this particular detail.

Reply:
I haven't personally had any trouble knowing when variables are updated atomically (especially when they're tightly scoped as here), but I'm not opposed to making it obvious. How about [this](https://play.golang.com/p/F3WMTYJA9aS)?

```go
package main

import (
	"fmt"
	"sync/atomic"
)

type AtomicUint32 struct{ uint32 }

func (a *AtomicUint32) Load() uint32 {
	return atomic.LoadUint32(&a.uint32)
}

func (a *AtomicUint32) Store(v uint32) {
	atomic.StoreUint32(&a.uint32, v)
}

func main() {
	var i AtomicUint32
	fmt.Println(i.Load())
	i.Store(12)
	fmt.Println(i.Load())
}

```

Reply:
> I haven't personally had any trouble knowing when variables are updated atomically (especially when they're tightly scoped as here), but I'm not opposed to making it obvious.

You're also a much more careful reader of code than most. 😉 

I personally find a naming convention to be the right balance between wordiness and obviousness, but ymmv. What you have is okay by me, too.